libclamav_rust/src/onenote.rs

```diff
@@ -0,0 +1,355 @@
+                // file appears to be OneStore 2.8 `.one` file.
+                section.page_series().iter().for_each(|page_series| {
+                    page_series.pages().iter().for_each(|page| {
+                        page.contents().iter().for_each(|page_content| {
+                            if let Some(page_outline) = page_content.outline() {
+                                page_outline.items().iter().for_each(|outline_item| {
+                                    outline_item.element().iter().for_each(|&outline_element| {
+                                        outline_element.contents().iter().for_each(|content| {
+                                            if let Some(embedded_file) = content.embedded_file() {
+                                                let data = embedded_file.data();
+                                                let name = embedded_file.filename();
+
+                                                // If name is empty, set to None.
+                                                let name = match name {
+                                                    "" => {
+                                                        debug!("Found unnamed attached file of size {}-bytes", data.len());
+                                                        None
+                                                    },
+                                                    _ => {
+                                                        debug!("Found attached file '{}' of size {}-bytes", name, data.len());
+                                                        Some(name.to_string())
+                                                    },
+                                                };
+
+                                                embedded_files.push(ExtractedFile {
+                                                    name: name,
+                                                    data: data.to_vec(),
+                                                });
+                                            }
+                                        });
+                                    });
+                                });
+                            }
+                        });
+                    });
+                });
+            } else {
+                return Err(OneDumpError::Parse);
+            }
+
+            Ok(embedded_files)
+        }
+
+        // Try to parse the section buffer using the onenote_parser crate.
+        // Attempt to catch panics in case the parser encounter unexpected issues.
+        let result_result = panic::catch_unwind(|| -> Result<Vec<ExtractedFile>, OneDumpError> {
+            let embedded_files = parse_section_buffer(data, filename)?;
+            Ok(embedded_files)
+        });
+
+        // Check if it panicked. If no panic, grab the parse result.
+        let result = match result_result {
```

Comment:
Love it.

Reply:
It gets even simpler if `OneDumpError::Parse` takes a value (as a tuple) that happens to be the error type you're mapping.  Then you can drop the closure syntax completely.

Reply:
I admit it would be nice to get a better explanation for a panic than just "Parse error" but the argument passed back is simple `Box<dyn Any + Send, Global>` and I don't know how to get meaningful information from that.

Reply:
Blame my inability to figure out the context within the [lousy] Github comment viewer.  I didn't realize that was the error from a panic trap.

> Unlike with normal errors, this value doesn’t implement the [Error](https://doc.rust-lang.org/std/error/trait.Error.html) trait.

Bleh...